pkgs/build-support/rust/fetchcargo.nix

```diff
@@ -21,9 +21,8 @@ stdenv.mkDerivation {
 
     export CARGO_HOME=$(mktemp -d cargo-home.XXX)
 
-    cargo vendor
-
-    cp -ar vendor $out
+    mkdir -p "$out"
+    cargo vendor "$out/sources" | sed "s|$out|@out@|g" > "$out/config"
```

Comment:
How stable is the output of this tool?
If we upgrade `cargo-vendor` we might break checksums again, right?

Reply:
We should probably ask @alexcrichton if we can agree on some kind of stable interface here.

Another possibility is to not include the content of config in the checksum but only the result of the final fetch. Not sure how this could be achived tough.

Reply:
The exact output to stdout probably isn't the most stable across versions of `cargo-vendor` but the output to the `vendor` directory should be deterministic.

Reply:
Can we replicate the output ourself instead? I don't want to updated all our checksum on every update of `cargo-vendor`

Reply:
I'm not sure how we'd do that without fragile Cargo.toml parsing. Perhaps @alexcrichton could comment on the feasibility of modifying cargo to make a static configuration viable?

Reply:
Sorry I'm not sure I understand the question :(

Reply:
The difficulty we're having here is that the cargo config varies according to the contents of the input `Cargo.toml`, and also apparently isn't stable enough to be relied upon, which means Nix has no reliable way to come up with the correct config file. If cargo could instead pick up git dependencies with magic similar to that which it uses for crates.io dependencies, the config file nix supplies could be static (as it is without this PR), and rust packages with git dependencies could be much more easily packaged.

Reply:
Hm I'm not sure I understand, isn't that expected? If a crate depends on git repositories we need to tell cargo where to find them instead, right?

Reply:
I don't have a particularly deep understanding of how cargo handles dependency resolution in general, but it seems strange that github dependencies must be explicitly enumerated (in an unstable fashion!) whereas crates.io ones can be auto-discovered.

Reply:
Cargo currently understands "sources" of packages and cargo-vendor emits a source-replacement configuration for each source necessary. All of crates.io is one source, while each git repository is another source.